### PR TITLE
feat: delete-zombies and clean to remove unneeded olean files

### PR DESCRIPTION
**Pull request by cipher1024**
TO CONTRIBUTORS:

Make sure you have:

  * [ ] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/naming.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/tests/tactics.lean)
  * [ ] make sure definitions and lemmas are put in the right files
  * [ ] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/code-review.md)

### CODE REVIEW

File: mathlibtools/leanproject.py [hunk 1]
```diff
@@ -205,6 +205,22 @@ def get_mathlib_cache(force: bool = False) -> None:
     except (LeanDownloadError, FileNotFoundError) as err:
         handle_exception(err, 'Failed to fetch mathlib oleans')
 
+@cli.command()
+def delete_zombies() -> None:
+    """Delete zombie oleans, .olean files with matching .lean files"""
```
Comment (urkud) [inline mathlibtools/leanproject.py]: "with no"?

File: mathlibtools/leanproject.py [hunk 2]
```diff
@@ -205,6 +205,22 @@ def get_mathlib_cache(force: bool = False) -> None:
...
+    try:
+        proj().delete_zombies()
+    except (LeanDownloadError, FileNotFoundError) as err:
+        handle_exception(err, 'Failed to delete zombies')
+
+@cli.command()
+def clean() -> None:
+    """Delete zombie oleans, .olean files with matching .lean files"""
```
Comment (urkud) [inline mathlibtools/leanproject.py]:
```suggestion
    """Delete all oleans"""
```

File: mathlibtools/leanproject.py [hunk 3]
```diff
@@ -205,6 +205,22 @@ def get_mathlib_cache(force: bool = False) -> None:
...
+        proj().delete_zombies()
+    except (LeanDownloadError, FileNotFoundError) as err:
+        handle_exception(err, 'Failed to delete zombies')
+
+@cli.command()
+def clean() -> None:
+    """Delete all olean files"""
+    try:
+        proj().clean()
+    except (LeanDownloadError, FileNotFoundError) as err:
```
Comment (PatrickMassot) [inline mathlibtools/leanproject.py]: Why do you want a specific error message for those two exceptions, and how LeanDownloadError could arise here?

File: mathlibtools/lib.py [hunk 4]
```diff
@@ -191,10 +191,36 @@ def lean_version_toml(version: VersionTuple) -> str:
         return ver_str
     else:
         return 'leanprover-community/lean:' + ver_str
-    
+
+
+def clean(dir):
+    print('cleaning {} ...'.format(dir))
+    for (path, _dirs, fs) in os.walk(dir):
```
Comment (PatrickMassot) [inline mathlibtools/lib.py]: This is very old style python. This project uses the pathlib standard library.

File: mathlibtools/lib.py [hunk 5]
```diff
@@ -191,10 +191,36 @@ def lean_version_toml(version: VersionTuple) -> str:
...
+        for fn in fs:
+            if os.path.splitext(fn)[1] == '.olean':
+                fn = Path(path, fn)
+                os.remove(fn)
+
+
+def delete_zombies(dir):
```
Comment (PatrickMassot) [inline mathlibtools/lib.py]: This seems very complicated for no reason. Did you have a look at how we check oleans are older than leans in the check function?

**Comment by PatrickMassot** [comment]
@cipher1024 I implemented the modifications I suggested in 7e9e82c , is that ok with you?

**Comment by cipher1024** [comment]
I appreciate it. Thanks for cleaning things up! Do you want us to change branch protections to make it easier for you to work on the tools?

**Pull request merged**